### DISCUSSION TITLE
Reactive function is called N times on Nth click

### DISCUSSION THREAD

**Discussion by akprasad** [alpinejs/alpine#3371] [7. Bugs (archived)] [answered]
Demo:
- https://jsfiddle.net/0a6dcf7x/

Instructions:
1. Open the JS console.
2. Repeatedly click "Click me" then "Close me"

Expected:
- `createWidgets` is printed once per click on `Click me`

Actual:
- `createWidgets` is printed once on the first click, twice on the second, and N times on the Nth click.

Comments:
- I notice that if I remove the `template x-if="true"` wrapper, the code works as expected. I'm guessing this is due to how Alpine's reactive data model interacts with x-if and x-show, but I'm not enough of an expert to know for sure.
- I tried searching for similar issues, but I'm not sure what terms to use in my search.

Thanks for any help you can provide.

**Comment by ekwoka** [comment, marked as answer]
As came up in another thread. the best way to do a conditional for loop is

```js
<template x-for="item in condition ? items : []">
```

Less elements 

But also, the issue is that you didn't call `createWidgets`. It should be `createWidgets()`

Make both of these changes has your demo working perfectly :D

Oh, and it should be a `button` not an `a`

**Reply by akprasad** [reply]
Hi @ekwoka,

Thanks very much for your reply and for the suggestions! I really appreciate that you took time to reply to me.

If I may, I have two follow-up questions:

1. In my actual application, there is some content that is within the `if` but before the `for` loop. You can see an example here -- look for `this is some sample content`: https://jsfiddle.net/12hLjyd7/3/ . Do you know how I can support this behavior? (Edit: I suppose I could have one `<template x-if...>` and one `<template x-for...>` (with the `?` syntax) as siblings to each other, though the repetition is unfortunate.)

2. The languages I'm used to (Python, Rust) will generally raise an error if I invoke a function incorrectly. So I'm quite surprised that my original demo still "works" rather than raising an error that something was called incorrectly. Does Alpine have a strict mode I can use during development to catch these errors sooner? Or if not, I would love to better understand the data model that caused the behavior I mentioned in my original post -- do you know of any docs I can read to understand why this "N times" behavior occurs?

**Reply by ekwoka** [reply]
It's more that it can't easily tell that what you've done is incorrect and not what you intended.

If your function returns a number you meant to return a boolean, nothing can really tell you you're wrong because it may think you meant to return a number.

But no Alpine has no stricter error reporting modes.

**Reply by akprasad** [reply]
Thanks, this resolves most of my questions so I'll mark this as the answer.

As a coda, I found that the error also goes away if I use `x-show` instead of `x-if`.

My totally speculative guess is that since `x-if` "completely adds and removes the element," Alpine is maintaining an extra reference to each usage of the `createWidgets()` function, including the usages in previous renders of the `x-if` block that have since been totally destroyed. As partial confirmation, using the same logic with `x-show` instead of `x-if` also removes the error entirely.

Going forward, I'll try to avoid `x-if` where possible.

**Reply by ekwoka** [reply]
Yes, it could be a bit of a bug that it's not doing a good job of cleaning up the effect, or somehow else duplicating it. I'd have to look at the logic for how it would be trying to evaluate the function as an iterator

**Reply by shreevatsa** [reply]
Note that doing just the second change (changing `createWidgets` to `createWidgets()`) doesn't seem to affect the bug or change anything; the unexpected behaviour still happens. So that does not seem to be the reason for the issue; is it just the `x-for` inside `x-if` that's causing the problem?

**Reply by ekwoka** [reply]
That should work as well unless something isn't cleaning up properly.

🤔

**Reply by shreevatsa** [reply]
Yeah here's something even more minimal that shows the same problem (if you look at the console logs): https://jsfiddle.net/mcLpoa2n/

**Reply by ekwoka** [reply]
looks like this could be a memory leak where the wrapping x-if is preventing the effects from being released when the outer x-if removes it.

it's the 2 x-ifs around it.
